crates/core/tedge_agent/src/entity_manager/tests.rs

```diff
@@ -0,0 +1,655 @@
+use crate::entity_manager::server::EntityStoreResponse;
+use crate::entity_manager::tests::model::Action;
+use crate::entity_manager::tests::model::Command;
+use crate::entity_manager::tests::model::Commands;
+use crate::entity_manager::tests::model::Protocol::MQTT;
+use proptest::proptest;
+use std::collections::HashSet;
+use tedge_actors::Server;
+use tedge_api::entity::EntityMetadata;
+
+#[tokio::test]
+async fn new_entity_store() {
+    let (mut entity_store, _mqtt_output) = entity::server("device-under-test");
+
+    assert_eq!(
+        entity::get(&mut entity_store, "device/main//").await,
+        Some(EntityMetadata::main_device())
+    )
+}
+
+#[tokio::test]
+async fn removing_an_unknown_child_using_mqtt() {
+    let registrations = vec![
+        // tedge mqtt pub -r te/device/a// ''
```

Comment:
What's the expected behavior of sending an empty MQTT message to an entity topic?

I was expecting this to delete the entity (which in this case has not been registered yet),
but this half register the entity:

- it is inserted in the `EntityTree` maintained by the entity store (using default metadata: i.e as a child device of main)
- but it is not listed as an entity (despite not being pending).

Reply:
> What's the expected behavior of sending an empty MQTT message to an entity topic?

The agent(entity store) will do nothing as deletions are supported only over HTTP. But the mapper would still react to that MQTT message by removing the that entry from its entity cache.

> but this half registers the entity: (using default metadata: i.e as a child device of main).

That's a bug in the auto-registration logic. A side effect of the entity manager actor of subscribing to `te/+/+/+/+/#` topic that includes `te/+/+/+/+` topic as well. It should have been `te/+/+/+/+/+/#` or to be completely on the safer side subscribe to the data topics specifically like `[te/+/+/+/+/m/+, te/+/+/+/+/e/+, te/+/+/+/+/a/+, te/+/+/+/+/cmd/+/+, ...]`.

Reply:
> That's a bug in the auto-registration logic.

Okay, I will fix that.

Reply:
Doing nothing on `tedge mqtt pub -r te/device/a// ''` is somehow inconsistent with the fact that an `HTTP DELETE /tedge/entity-store/v1/entities/device/a//` publish a retain empty message on `te/device/a//`. How can a client like the mapper tell the difference between the two cases?

It would be better for the agent to react on clearing an entity topic by deleting the entity.

Reply:
> It would be better for the agent to react on clearing an entity topic by deleting the entity.

I agree. Now that `tedge-agent` is the sole owner of the entity store, we can treat empty MQTT messages also as a request to delete the entire tree.

Reply:
>> It would be better for the agent to react on clearing an entity topic by deleting the entity.

> I agree. Now that `tedge-agent` is the sole owner of the entity store, we can treat empty MQTT messages also as a request to delete the entire tree.

Implemented by https://github.com/thin-edge/thin-edge.io/pull/3344/commits/5009140e9b8980ef68f5b38ebd7e9bebdb417698